// @sanity/desk-tool/lib/components/ConfirmPublish.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _fullscreen = require('part:@sanity/components/dialogs/fullscreen');

var _fullscreen2 = _interopRequireDefault(_fullscreen);

var _distance_in_words_to_now = require('date-fns/distance_in_words_to_now');

var _distance_in_words_to_now2 = _interopRequireDefault(_distance_in_words_to_now);

var _DocTitle = require('./DocTitle');

var _DocTitle2 = _interopRequireDefault(_DocTitle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/no-multi-comp */
const ACTIONS = [{ name: 'confirm', title: 'Publish now', color: 'success', autoFocus: true, inverted: false }, { name: 'cancel', title: 'Cancel' }];
const NOTHING_TO_PUBLISH_ACTIONS = [{ name: 'cancel', title: 'Close', color: 'success', autoFocus: true }];

class ConfirmPublish extends _react2.default.PureComponent {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.handleAction = action => {
      var _props = this.props;
      const onCancel = _props.onCancel,
            onConfirm = _props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }
      if (action.name === 'cancel') {
        onCancel();
      }
    }, _temp;
  }

  render() {
    var _props2 = this.props;
    const draft = _props2.draft,
          published = _props2.published,
          onCancel = _props2.onCancel;


    if (!draft) {
      return _react2.default.createElement(
        _fullscreen2.default,
        {
          isOpen: true,
          centered: true,
          showHeader: true,
          title: 'No changes',
          actions: NOTHING_TO_PUBLISH_ACTIONS,
          onAction: this.handleAction
        },
        'Nothing to publish!'
      );
    }
    return _react2.default.createElement(
      _fullscreen2.default,
      {
        isOpen: true,
        showHeader: true,
        title: 'Confirm publish',
        centered: true,
        onClose: onCancel,
        onAction: this.handleAction,
        actions: ACTIONS
      },
      _react2.default.createElement(
        'div',
        { style: { padding: 10 } },
        _react2.default.createElement(
          'p',
          null,
          'Are you sure you would like to publish the document',
          ' ',
          _react2.default.createElement(
            'strong',
            null,
            _react2.default.createElement(_DocTitle2.default, { document: draft || published })
          ),
          '?'
        ),
        _react2.default.createElement(
          'p',
          null,
          published && `It was last published ${(0, _distance_in_words_to_now2.default)(published._updatedAt, { addSuffix: true })}.`
        )
      )
    );
  }
}
exports.default = ConfirmPublish;
ConfirmPublish.propTypes = {
  onCancel: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func.isRequired,
  published: _propTypes2.default.object,
  draft: _propTypes2.default.object
};
