// @sanity/desk-tool/lib/components/ConfirmDiscard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _fullscreen = require('part:@sanity/components/dialogs/fullscreen');

var _fullscreen2 = _interopRequireDefault(_fullscreen);

var _DocTitle = require('./DocTitle');

var _DocTitle2 = _interopRequireDefault(_DocTitle);

var _distance_in_words_to_now = require('date-fns/distance_in_words_to_now');

var _distance_in_words_to_now2 = _interopRequireDefault(_distance_in_words_to_now);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/no-multi-comp */
const CANCEL_ACTION = { name: 'cancel', title: 'Cancel', kind: 'secondary' };

class ConfirmDiscard extends _react2.default.PureComponent {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.handleAction = action => {
      var _props = this.props;
      const onCancel = _props.onCancel,
            onConfirm = _props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }
      if (action.name === 'cancel') {
        onCancel();
      }
    }, _temp;
  }

  render() {
    var _props2 = this.props;
    const draft = _props2.draft,
          published = _props2.published,
          onCancel = _props2.onCancel;

    const confirmAction = {
      name: 'confirm',
      title: `Discard ${published ? 'changes' : 'draft'} now`,
      color: 'danger'
    };
    return _react2.default.createElement(
      _fullscreen2.default,
      {
        isOpen: true,
        showHeader: true,
        title: `Confirm discard ${published ? 'changes' : 'document'}`,
        centered: true,
        onClose: onCancel,
        onAction: this.handleAction,
        actions: [confirmAction, CANCEL_ACTION]
      },
      _react2.default.createElement(
        'div',
        { style: { padding: 10 } },
        _react2.default.createElement(
          'p',
          null,
          'Are you sure you would like to discard ',
          published ? 'changes in' : 'the document',
          ' ',
          _react2.default.createElement(
            'strong',
            null,
            _react2.default.createElement(_DocTitle2.default, { document: draft || published })
          ),
          '?'
        ),
        _react2.default.createElement(
          'p',
          null,
          'This will ',
          published ? `revert to the latest published version of this document (published ${(0, _distance_in_words_to_now2.default)(published._updatedAt, { addSuffix: true })}).` : 'delete it entirely and there is no going back.'
        )
      )
    );
  }
}
exports.default = ConfirmDiscard;
ConfirmDiscard.propTypes = {
  onCancel: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func.isRequired,
  published: _propTypes2.default.object,
  draft: _propTypes2.default.object
};
