// @sanity/desk-tool/lib/components/enhanceWithReferringDocuments.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = withReferringDocuments;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _document = require('part:@sanity/base/datastore/document');

var _document2 = _interopRequireDefault(_document);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function withReferringDocuments(Component) {
  var _class, _temp2;

  return _temp2 = _class = class extends _react2.default.PureComponent {
    constructor(...args) {
      var _temp;

      return _temp = super(...args), this.state = {
        isLoading: false,
        referringDocuments: []
      }, _temp;
    }

    componentDidMount() {
      const published = this.props.published;

      if (!published) {
        return;
      }
      this.setState({ isLoading: true });
      this.refSubscription = _document2.default.query('*[references($docId)] [0...101]', { docId: published._id }).subscribe(event => {
        this.setState({
          referringDocuments: event.documents || [],
          isLoading: false
        });
      });
    }

    componentWillUnmount() {
      if (this.refSubscription) {
        this.refSubscription.unsubscribe();
      }
    }

    render() {
      var _state = this.state;
      const isLoading = _state.isLoading,
            referringDocuments = _state.referringDocuments;

      return _react2.default.createElement(Component, _extends({}, this.props, {
        referringDocuments: referringDocuments,
        isCheckingReferringDocuments: isLoading
      }));
    }
  }, _class.displayName = `enhanceWithAvailHeight(${Component.displayName || Component.name})`, _class.propTypes = {
    published: _propTypes2.default.object
  }, _temp2;
}
