// @sanity/desk-tool/lib/components/TimeAgo.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _shallowEquals = require('shallow-equals');

var _shallowEquals2 = _interopRequireDefault(_shallowEquals);

var _distance_in_words_to_now = require('date-fns/distance_in_words_to_now');

var _distance_in_words_to_now2 = _interopRequireDefault(_distance_in_words_to_now);

var _format = require('date-fns/format');

var _format2 = _interopRequireDefault(_format);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class TimeAgo extends _react2.default.PureComponent {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.update = () => {
      this.forceUpdate();
    }, _temp;
  }

  componentDidMount() {
    this.start();
  }

  componentWillUnmount() {
    this.stop();
  }

  componentWillReceiveProps(nextProps) {
    if (!(0, _shallowEquals2.default)(this.props, nextProps)) {
      this.stop();
      this.start();
    }
  }

  start() {
    this.update();
    this.intervalId = setInterval(this.update, this.props.refreshInterval);
  }

  stop() {
    clearInterval(this.intervalId);
  }

  render() {
    const timestamp = (0, _format2.default)(this.props.time, 'MMM D, YYYY, h:mm A Z');
    return _react2.default.createElement(
      'span',
      { title: timestamp },
      (0, _distance_in_words_to_now2.default)(this.props.time, { addSuffix: true })
    );
  }
}
exports.default = TimeAgo;
TimeAgo.propTypes = {
  refreshInterval: _propTypes2.default.number,
  time: _propTypes2.default.string.isRequired
};
TimeAgo.defaultProps = {
  refreshInterval: 1000 * 30
};
