// @sanity/desk-tool/lib/pane/TypePaneItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TypePaneItem = require('./styles/TypePaneItem.css');

var _TypePaneItem2 = _interopRequireDefault(_TypePaneItem);

var _router = require('part:@sanity/base/router');

var _reactInk = require('react-ink');

var _reactInk2 = _interopRequireDefault(_reactInk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class TypePaneItem extends _react2.default.Component {

  render() {
    var _props = this.props;
    const selected = _props.selected,
          type = _props.type;


    return _react2.default.createElement(
      'div',
      {
        className: selected ? _TypePaneItem2.default.selected : _TypePaneItem2.default.item,
        key: document._id
      },
      _react2.default.createElement(
        _router.StateLink,
        {
          state: { selectedType: type.name },
          className: _TypePaneItem2.default.link
        },
        type.title,
        _react2.default.createElement(_reactInk2.default, { duration: 1000, opacity: 0.1, radius: 200 })
      )
    );
  }
}
exports.default = TypePaneItem;
TypePaneItem.propTypes = {
  type: _propTypes2.default.shape({
    name: _propTypes2.default.string,
    title: _propTypes2.default.string,
    icon: _propTypes2.default.element
  }),
  selected: _propTypes2.default.bool
};
