// @sanity/components/lib/progress/ProgressCircle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _circleStyle = require('part:@sanity/components/progress/circle-style');

var _circleStyle2 = _interopRequireDefault(_circleStyle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const radiusFactor = 1.3;
const widthFactor = 1.2;
const heightFactor = 1.2;

class ProgressCircle extends _react2.default.Component {

  render() {
    var _props = this.props;
    const percent = _props.percent,
          completed = _props.completed,
          text = _props.text,
          style = _props.style,
          showPercent = _props.showPercent,
          radius = _props.radius,
          strokeWidth = _props.strokeWidth;

    const rootClasses = `
      ${completed ? _circleStyle2.default.completed : _circleStyle2.default.uncompleted}
      ${percent >= 100 && _circleStyle2.default.hundredPercent}
    `;

    const width = radius * 2;
    const height = radius * 2;
    const viewBox = `-10 -10 ${width * widthFactor} ${height * heightFactor}`;

    const dashArray = this.props.radius * Math.PI * 2;
    const dashOffset = dashArray - dashArray * percent / 100; //eslint-disable-line no-mixed-operators

    return _react2.default.createElement(
      'div',
      { className: rootClasses, style: style },
      _react2.default.createElement(
        'div',
        { className: _circleStyle2.default.inner },
        _react2.default.createElement(
          'svg',
          {
            className: _circleStyle2.default.svg,
            width: width,
            height: height,
            viewBox: viewBox
          },
          _react2.default.createElement('circle', {
            className: _circleStyle2.default.background,
            cx: radius,
            cy: radius,
            r: radius,
            strokeWidth: `${strokeWidth}px`
          }),
          _react2.default.createElement('circle', {
            className: _circleStyle2.default.foreground,
            cx: radius,
            cy: radius,
            r: radius,
            strokeWidth: `${strokeWidth}px`,
            style: {
              strokeDasharray: dashArray,
              strokeDashoffset: dashOffset
            }
          }),
          showPercent && _react2.default.createElement(
            'text',
            {
              className: _circleStyle2.default.percent,
              x: radius,
              y: text ? radius - 5 : radius,
              dy: '.4em',
              textAnchor: 'middle'
            },
            `${Math.round(percent, 1)}%`
          ),
          text && _react2.default.createElement(
            'text',
            {
              className: _circleStyle2.default.status,
              x: radius,
              y: radius * radiusFactor - 5,
              dy: '.4em',
              textAnchor: 'middle'
            },
            text
          )
        )
      )
    );
  }
}
exports.default = ProgressCircle;
ProgressCircle.propTypes = {
  className: _propTypes2.default.string,
  percent: _propTypes2.default.number,
  animation: _propTypes2.default.bool,
  text: _propTypes2.default.string,
  style: _propTypes2.default.object,
  showPercent: _propTypes2.default.bool,
  radius: _propTypes2.default.number,
  strokeWidth: _propTypes2.default.number,
  status: _propTypes2.default.string,
  completed: _propTypes2.default.bool
};
ProgressCircle.defaultProps = {
  radius: 50,
  percentage: 0,
  strokeWidth: 10
};
