// @sanity/base/lib/datastores/document/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _observable = require('@sanity/observable');

var _observable2 = _interopRequireDefault(_observable);

var _documentStore = require('@sanity/document-store');

var _documentStore2 = _interopRequireDefault(_documentStore);

var _client = require('part:@sanity/base/client');

var _client2 = _interopRequireDefault(_client);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function fetchDocumentSnapshot(id) {
  return _client2.default.observable.getDocument(id).map(document => {
    return {
      type: 'snapshot',
      document: document
    };
  });
}

function fetchQuerySnapshot(query, params) {
  return _client2.default.observable.fetch(query, params).map(documents => {
    return {
      type: 'snapshot',
      documents: documents
    };
  });
}

const serverConnection = {
  byId(id) {
    return _observable2.default.from(_client2.default.listen('*[_id == $id]', { id: id }, { includeResult: false, events: ['welcome', 'mutation', 'reconnect'] })).concatMap(event => {
      return event.type === 'welcome' ? _observable2.default.from(fetchDocumentSnapshot(id)) : _observable2.default.of(event);
    });
  },

  query(query, params) {
    return _observable2.default.from(_client2.default.observable.listen(query, params || {}, { includeResult: false, events: ['welcome', 'mutation', 'reconnect'] })).concatMap(event => {
      return event.type === 'welcome' ? _observable2.default.from(fetchQuerySnapshot(query, params)) : _observable2.default.of(event);
    });
  },

  mutate(mutations) {
    return _observable2.default.from(_client2.default.observable.dataRequest('mutate', mutations, { visibility: 'async', returnDocuments: false }));
  },

  delete(id) {
    return _observable2.default.from(_client2.default.observable.delete(id, { visibility: 'async', returnDocuments: false }));
  },

  create(doc) {
    return _observable2.default.from(_client2.default.observable.create(doc));
  },
  createIfNotExists(doc) {
    return _observable2.default.from(_client2.default.observable.createIfNotExists(doc));
  },
  createOrReplace(doc) {
    return _observable2.default.from(_client2.default.observable.createOrReplace(doc));
  }
};

exports.default = (0, _documentStore2.default)({ serverConnection });
